src/nunavut/templates.py

```diff
@@ -0,0 +1,120 @@
+#
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# Copyright (C) 2018-2019  UAVCAN Development Team  <uavcan.org>
+# This software is distributed under the terms of the MIT License.
+#
+"""
+Abstractions around template engine internals.
+"""
+import typing
+
+EnvironmentFilterAttributeName = 'environmentfilter'
+"""
+For now this is set to a value that is internally compatible with the
+embedded version of jinja2 we use in Nunavut. If you use this variable
+instead of its current value you will be insulated from this.
+"""
+
+ContextFilterAttributeName = 'contextfilter'
+"""
+For now this is set to a value that is internally compatible with the
+embedded version of jinja2 we use in Nunavut. If you use this variable
+instead of its current value you will be insulated from this.
+"""
+
+LanguageFilterAttributeName = 'nv_languagefilter'
+"""
+Nunavut-specific attribute for filters that take their :class:`nunavut.lang.Language`
+as the first argument.
+"""
+
+
+class SupportsTemplateEnv:
+    """
+    Provided as a pseudo
+    `protocol <https://mypy.readthedocs.io/en/latest/protocols.html#simple-user-defined-protocols/>`_.
+    (in anticipation of that becoming part of the core Python typing someday).
+
+    :var Dict globals: A dictionary mapping global names (str) to variables (Any) that are available in
+                       each template's global environment.
+    :var Dict filters: A dictionary mapping filter names (str) to filters (Callable). Filters are simply
+                       global functions available to templates but are most often used to transform a
+                       given input to output emitted by the template.
+    """
+
+    globals = dict()  # type: typing.Dict[str, typing.Any]
+    filters = dict()  # type: typing.Dict[str, typing.Callable]
+    tests = dict()  # type: typing.Dict[str, typing.Callable[[typing.Any], bool]]
+
+
+class SupportsTemplateContext:
+    """
+    Provided as a pseudo
+    `protocol <https://mypy.readthedocs.io/en/latest/protocols.html#simple-user-defined-protocols/>`_.
+    (in anticipation of that becoming part of the core Python typing someday).
+    """
+
+
+def templateEnvironmentFilter(filter_func: typing.Callable) -> typing.Callable[..., str]:
```

Comment:
Per PEP8, this should be named in `snake_case`. Why is it different here?

Reply:
I believe decorators are in camel case for PEP8?

Reply:
PEP8 says this:

> mixedCase is allowed only in contexts where that's already the prevailing style (e.g. threading.py), to retain backwards compatibility.

Decorators are generally functions so they should be in snake_case.

Reply:
I'd like to keep this camelcase to be more consistent with JINJA decorators.

Reply:
![image](https://user-images.githubusercontent.com/19180861/70652528-9944e780-1c07-11ea-9b90-781015dbe8c5.png)